### PR TITLE
envoy: Update to release 1.21.0

### PR DESCRIPTION
**Pull request by jrajahalme**
Update Envoy to release 1.21.0. Envoy Go API is updated to contain
the generated validation code.

cilium-envoy image is updated to support the newEgressMarkSourceEndpointId
option for the Cilium listener filter. NPDS field 'Policy' is renamed as
'EndpointID'. 'Policy' field was not used for anything, so might as
well recycle it while this API is not yet public.

Envoy retries may fail on "address already in use" when the original
source address and port are used on upstream connections. Cilium
typically does this in the egress proxy listeners. Fix this by using a
Cilium Envoy build that always sets SO_REUSEADDR when original source
address and port is used.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

```release-note
Cilium host proxy is updated to Envoy release 1.21.0
```

### CODE REVIEW

**Review by rolinh — APPROVED** [review]

**Comment by jrajahalme** [comment]
/test-only  --focus="K8sPolicyTest.*using namespace label and L7"

**Comment by jrajahalme** [comment]
/test-only --focus="K8sPolicyTest.*using.namespace.label.and.L7"

**Comment by jrajahalme** [comment]
Focused test run "failed" due to artifact collection failing when no tests were run:
```
09:59:19  Ran 0 of 401 Specs in 9.062 seconds
09:59:19  SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 401 Skipped
09:59:19  PASS
```

**Comment by jrajahalme** [comment]
Fixed typo.

**Comment by jrajahalme** [comment]
runtime test failed with unrelated error:
```
14:44:28      runtime: /tmp/provision/compile.sh: line 28: cd: /home/vagrant/go/src/github.com/cilium/cilium: Stale file handle
```
This likely indicates a problem in the VM provisioning.

**Review by michi-covalent — APPROVED** [review]

File: pkg/envoy/server_test.go [hunk 1]
```diff
@@ -442,7 +442,7 @@ func (s *ServerSuite) TestGetNetworkPolicy(c *C) {
 	obtained := getNetworkPolicy(ep, nil, IPv4Addr, L4Policy1, true, true)
 	expected := &cilium.NetworkPolicy{
 		Name:                   IPv4Addr,
-		Policy:                 uint64(Identity),
```
Comment (sayboras) [inline pkg/envoy/server_test.go:445]: nit: uint64 conversion here is redundant

**Review by sayboras — APPROVED** [review]
LGTM. 

multi cluster test can be re-run after rebase with master.

File: daemon/cmd/policy_test.go [hunk 2]
```diff
@@ -301,19 +301,19 @@ func (ds *DaemonSuite) TestUpdateConsumerMap(c *C) {
...
+	c.Assert(eQABar.Allows(qaFooSecLblsCtx.ID), Equals, true)
+	c.Assert(eQABar.Allows(prodFooSecLblsCtx.ID), Equals, false)
+
+	eProdBar := ds.prepareEndpoint(c, prodBarSecLblsCtx, false)
+	c.Assert(eProdBar.Allows(0), Equals, false)
+	c.Assert(eProdBar.Allows(qaBarSecLblsCtx.ID), Equals, false)
+	c.Assert(eProdBar.Allows(prodBarSecLblsCtx.ID), Equals, false)
+	c.Assert(eProdBar.Allows(qaFooSecLblsCtx.ID), Equals, false)
+	c.Assert(eProdBar.Allows(prodFooSecLblsCtx.ID), Equals, true)
+	c.Assert(eProdBar.Allows(prodFooJoeSecLblsCtx.ID), Equals, true)
```
Comment (joestringer) [inline daemon/cmd/policy_test.go:316]: What do these changes have to do with Envoy update?

Comment (joestringer) [reply daemon/cmd/policy_test.go:316]: (I'm OK with the changes, they just seem like extra churn for something we intend to backport so maybe we should split them into a new master-only PR?)

File: pkg/envoy/server.go [hunk 3]
```diff
@@ -479,7 +479,14 @@ func (s *XDSServer) addListener(name string, port uint16, listenerConf func() *e
 	}
 	listener.mutex.Unlock() // Listener locked again in callbacks below
 
-	s.listenerMutator.Upsert(ListenerTypeURL, name, listenerConf(), []string{"127.0.0.1"}, wg,
+	listenerConfig := listenerConf()
+	if option.Config.EnableBPFTProxy {
+		// Envoy since 1.20.0 uses SO_REUSEPORT on listeners by default.
+		// BPF TPROXY is currently not compatible with SO_REUSEPORT, so disable it.
+		// Note that this may degrade Envoy performance.
+		listenerConfig.EnableReusePort = &wrappers.BoolValue{Value: false}
```
Comment (joestringer) [inline pkg/envoy/server.go:487]: :+1: Ack

File: daemon/cmd/policy_test.go [hunk 4]
```diff
@@ -377,7 +377,7 @@ func (ds *DaemonSuite) TestUpdateConsumerMap(c *C) {
 
 	expectedNetworkPolicy = &cilium.NetworkPolicy{
 		Name:             ProdIPv4Addr.String(),
-		Policy:           uint64(prodBarSecLblsCtx.ID),
+		EndpointId:       uint64(eProdBar.ID),
```
Comment (joestringer) [inline daemon/cmd/policy_test.go:380]: Interesting, so the policy API between cilium-agent and the proxy has changed?

I think this is fine given that we always ship Envoy alongside Cilium. So we shouldn't need to deal with compatibility issues of mismatched versions attempting to use different field names from what the other binary expects.

**Review by joestringer — APPROVED** [review]
LGTM.

File: daemon/cmd/policy_test.go (hunk 2 above)
Comment (joestringer) [reply daemon/cmd/policy_test.go:316]: ^^ Probably not worth changing, this rename makes the tests a bit more readable and prevents potential conflicts with variables with a simple name like `e`. Let's just :ship:

**Comment by joestringer** [comment]
Multicluster workflow was broken on master for a while, fixed in https://github.com/cilium/cilium/pull/18808 (merged yesterday). It's likely only failing because the PR has not been rebased. I think we're good to optimistically just merge this to unblock next steps.

**Pull request merged**